### PR TITLE
Reserve more numeric types.

### PR DESCRIPTION
**Pull request by mahkoh**
[Rendered](https://github.com/mahkoh/rfcs/blob/primitives/text/0000-reserve-types.md)

### CODE REVIEW

**Comment by mahkoh** [comment]
Previous discussion: https://github.com/rust-lang/rfcs/pull/138

**Comment by huonw** [comment]
What is this protecting against? That is, what's a program that will stop compiling if we don't reserve them?

It seems to me (and the filer of the original RFC) that shadowing should mean things work fine.

**Comment by mahkoh** [comment]
>That is, what's a program that will stop compiling if we don't reserve them?

As mentioned in the RFC:

```
<anon>:1:1: 1:15 error: user-defined types or type parameters cannot shadow the primitive types [E0317]
<anon>:1 type u8 = u16;
         ^~~~~~~~~~~~~~
```

File: text/0000-reserve-types.md [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+It is conceivable that Rust will gain support for types such as `f128`, `f16`,
+or `u128`. In the interest of backwards compatability, extend the grammar to
+reserve these.
+
+Adding them is backwards incompatible because `type PRIMITIVE = T;` is an error.
+
+# Detailed design
+
+Reserve the following type names: `fN`, `uN`, `iN` for `N` a multiple of 8.
```
Comment (huonw) [inline text/0000-reserve-types.md]: What limit?

Comment (mahkoh) [inline text/0000-reserve-types.md]: Whatever the parser can handle.

Comment (huonw) [inline text/0000-reserve-types.md]: The parser can handle arbitrarily large numbers: congruence mod 8 can be determined using only the last 3 digits of a number represented in base 10.

**Comment by Thiez** [comment]
Why take only multiples of 8 for `N`? If this change is going to be made we might as well take all of them, LLVM has some [support](http://llvm.org/docs/LangRef.html#integer-type) for this.

**Comment by cmr** [comment]
That's a pretty annoying shadowing restriction right there, though it does avoid exposing the resolve bugs around such things...

**Comment by petrochenkov** [comment]
@cmr 
This restriction is temporary and will be removed if @eddyb 's plan works
See https://github.com/rust-lang/rust/issues/20427#issuecomment-74576273

File: text/0000-reserve-types.md (hunk 1 above)
Comment (jmesmon) [inline text/0000-reserve-types.md]: Is there any reason to limit ourselves to numbers that are a multiple of 8? Why not just reserve them for all numbers?

**Comment by eddyb** [comment]
:-1: to this RFC as written, it's a really large hammer and only makes sense if we had arbitrarily sized primitives (like `iN` in LLVM, but we don't expose that - not that backends can deal with uncommon bit widths, anyways).

The only approach I know of and I consider (more or less) "correct" would be the one mentioned in the commit linked by @petrochenkov above.

That would protect against breakage by explicitly having all existing built-in types part of `prelude::v1`, which means `f128` can be added backwards-compatibly, by requiring it to be manually imported from `std::num` (and only show up in `prelude::v2`, if ever).

That said, I almost forgot about it, and it requires associated constants to be done backwards-compatibly (or as close as we can get to that).

**Comment by bombless** [comment]
We should fix the bug itself, not to shadow it with strange behavior.

**Comment by alexcrichton** [comment]
Thanks for the RFC! After some discussion, however, we don't think that this is the strategy that we would want to take. It is arguably a resolve bug which prevents shadowing primitives and we can consider it a requirement that adding a type like `f16` would require fixing the resolve bug first to allow shadowing of primitives. As such the change should still be backwards compatible, just perhaps a little harder than before, so I'm going to close this.

**Pull request closed without merging**